rust/lance/src/datatypes/schema.rs

```diff
@@ -35,6 +35,60 @@ pub struct Schema {
     pub metadata: HashMap<String, String>,
 }
 
+/// State for a pre-order DFS iterator over the fields of a schema.
+struct SchemaFieldIterPreOrder<'a> {
+    schema: &'a Schema,
+    field_stack: Vec<&'a Field>,
+    idx_stack: Vec<usize>,
+    top_level_idx: usize,
```

Comment:
bikeshed: Would be simpler just to have 1 stack: insert fields in reverse order, pop from the back, if nested push children back in reverse order?

Reply:
That was a lot simpler.  Great idea!